src/mscorlib/src/System/String.cs

```diff
@@ -465,88 +458,122 @@ private unsafe static bool StartsWithOrdinalHelper(String str, String startsWith
         [System.Security.SecuritySafeCritical]  // auto-generated
         private unsafe static int CompareOrdinalHelper(String strA, String strB)
         {
-            Contract.Requires(strA != null);
-            Contract.Requires(strB != null);
+            Contract.Assert(strA != null);
+            Contract.Assert(strB != null);
+
+            // NOTE: This may be subject to change if eliminating the check
+            // in the callers makes them small enough to be inlined by the JIT
+            Contract.Assert(strA.m_firstChar == strB.m_firstChar,
+                "For performance reasons, callers of this method should " +
+                "check/short-circuit beforehand if the first char is the same.");
 
             int length = Math.Min(strA.Length, strB.Length);
-            int diffOffset = -1;
 
             fixed (char* ap = &strA.m_firstChar) fixed (char* bp = &strB.m_firstChar)
             {
                 char* a = ap;
                 char* b = bp;
 
+                // Check if the second chars are different here
+                // The reason we check if m_firstChar is different is because
+                // it's the most common case and allows us to avoid a method call
+                // to here.
+                // The reason we check if the second char is different is because
+                // if the first two chars the same we can increment by 4 bytes,
+                // leaving us word-aligned on both 32-bit (12 bytes into the string)
+                // and 64-bit (16 bytes) platforms.
+        
+                // For empty strings, the second char will be null due to padding.
+                // The start of the string (not including sync block pointer)
+                // is the method table pointer + string length, which takes up
+                // 8 bytes on 32-bit, 12 on x64. For empty strings the null
+                // terminator immediately follows, leaving us with an object
+                // 10/14 bytes in size. Since everything needs to be a multiple
+                // of 4/8, this will get padded and zeroed out.
+                
+                // For one-char strings the second char will be the null terminator.
+
+                // NOTE: If in the future there is a way to read the second char
+                // without pinning the string (e.g. System.Runtime.CompilerServices.Unsafe
+                // is exposed to mscorlib, or a future version of C# allows inline IL),
+                // then do that and short-circuit before the fixed.
+
+                if (*(a + 1) != *(b + 1))
+                {
+                    return *(a + 1) - *(b + 1);
```

Comment:
@jkotas OK, makes sense. Regarding the conditional, do you think it would be better to change this check to

```cs
if (*(int*)a != *(int*)b)
```

or

```cs
if (*a++ != *b++) goto DiffNextChar;
```

I'm asking because I'm not sure if there's any overhead of adding to a pointer before deferencing it; in the [generated assembly](https://gist.github.com/jamesqo/de2fa42f6ffe30e1bbabf3e893051320#file-compareordinalhelper-x86-asm-L65) I see things like `movzx    eax, word  ptr [esi+2]`. Does adding 2 to the pointer make it slower than dereferencing `word  ptr [esi]` directly? (Forgive my inexperience with assembly)